File: backend/services/webhook.ts
import crypto from 'crypto';
import type {
  WebhookAnalytics,
  WebhookConfig,
  WebhookDelivery,
  WebhookDeliveryStatus,
  WebhookEventInput,
  WebhookEventPayload,
  WebhookEventType,
  WebhookRetryPolicy,
} from '../../src/types/webhook';

export type { WebhookEventInput } from '../../src/types/webhook';

type FetchLike = typeof fetch;

export interface RegisterWebhookInput {
  merchantId: string;
  url: string;
  events: WebhookEventType[];
  secretKey: string;
  retryPolicy?: Partial<WebhookRetryPolicy>;
  isPaused?: boolean;
}

export interface WebhookDeliveryResult {
  delivery: WebhookDelivery;
  response?: Response;
}

const MAX_PAYLOAD_BYTES = 1_048_576;
const DEFAULT_RETRY_POLICY: WebhookRetryPolicy = {
  maxRetries: 5,
  initialDelayMs: 250,
  maxDelayMs: 8_000,
  backoffFactor: 2,
};

const now = (): number => Date.now();

const createId = (prefix: string): string =>
  `${prefix}_${now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`;

const clampRetryPolicy = (retryPolicy?: Partial<WebhookRetryPolicy>): WebhookRetryPolicy => ({
  maxRetries: retryPolicy?.maxRetries ?? DEFAULT_RETRY_POLICY.maxRetries,
  initialDelayMs: retryPolicy?.initialDelayMs ?? DEFAULT_RETRY_POLICY.initialDelayMs,
  maxDelayMs: retryPolicy?.maxDelayMs ?? DEFAULT_RETRY_POLICY.maxDelayMs,
  backoffFactor: retryPolicy?.backoffFactor ?? DEFAULT_RETRY_POLICY.backoffFactor,
});

const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

export const signWebhookPayload = (payload: WebhookEventPayload, secretKey: string): string => {
  const body = JSON.stringify(payload);
  return crypto.createHmac('sha256', secretKey).update(body).digest('hex');
};

export const verifyWebhookSignature = (
  signature: string,
  payload: WebhookEventPayload,
  secretKey: string
): boolean => {
  const expected = signWebhookPayload(payload, secretKey);
  const actualBytes = Buffer.from(signature);
  const expectedBytes = Buffer.from(expected);
  if (actualBytes.length !== expectedBytes.length) return false;
  return crypto.timingSafeEqual(actualBytes, expectedBytes);
};

export const buildWebhookPayload = (input: WebhookEventInput): WebhookEventPayload => {
  const eventId = createId('evt');
  return {
    id: eventId,
    webhookId: input.webhookId,
    eventType: input.eventType,
    occurredAt: input.occurredAt ?? now(),
    merchantId: input.merchantId,
    subscription: input.subscription,
    plan: input.plan,
    previousStatus: input.previousStatus,
    currentStatus: input.currentStatus,
    payloadVersion: 1,
  };
};

export const isWebhookEventAllowed = (
  webhook: Pick<WebhookConfig, 'events' | 'isPaused'>,
  eventType: WebhookEventType
): boolean => !webhook.isPaused && webhook.events.includes(eventType);

export class WebhookDeliveryService {
  private readonly fetchImpl: FetchLike;
  private readonly sleepImpl: (ms: number) => Promise<void>;
  private readonly webhooks = new Map<string, WebhookConfig>();
  private readonly deliveries = new Map<string, WebhookDelivery>();
  private readonly deliveredKeys = new Set<string>();

  constructor(options: { fetchImpl?: FetchLike; sleepImpl?: (ms: number) => Promise<void> } = {}) {
    this.fetchImpl = options.fetchImpl ?? fetch;
    this.sleepImpl = options.sleepImpl ?? sleep;
  }

  registerWebhook(input: RegisterWebhookInput): WebhookConfig {
    const id = createId('whk');
    const createdAt = now();
    const config: WebhookConfig = {
      id,
      merchantId: input.merchantId,
      url: input.url,
      events: [...input.events],
      secretKey: input.secretKey,
      retryPolicy: clampRetryPolicy(input.retryPolicy),
      isPaused: input.isPaused ?? false,
      createdAt,
      updatedAt: createdAt,
      lastHealthCheckAt: undefined,
      lastHealthStatus: undefined,
      successCount: 0,
      failureCount: 0,
    };

    this.webhooks.set(id, config);
    return config;
  }

  updateWebhook(
    id: string,
    input: Partial<Omit<RegisterWebhookInput, 'merchantId'>>
  ): WebhookConfig {
    const existing = this.webhooks.get(id);
    if (!existing) throw new Error(`Webhook ${id} not found`);

    const next: WebhookConfig = {
      ...existing,
      url: input.url ?? existing.url,
      events: input.events ? [...input.events] : existing.events,
      secretKey: input.secretKey ?? existing.secretKey,
      retryPolicy: clampRetryPolicy(input.retryPolicy ?? existing.retryPolicy),
      isPaused: input.isPaused ?? existing.isPaused,
      updatedAt: now(),
    };

    this.webhooks.set(id, next);
    return next;
  }

  deleteWebhook(id: string): void {
    this.webhooks.delete(id);
  }

  pauseWebhook(id: string): WebhookConfig {
    return this.updateWebhook(id, { isPaused: true });
  }

  resumeWebhook(id: string): WebhookConfig {
    return this.updateWebhook(id, { isPaused: false });
  }

  listWebhooks(merchantId: string): WebhookConfig[] {
    return Array.from(this.webhooks.values()).filter(
      (webhook) => webhook.merchantId === merchantId
    );
  }

  getWebhook(id: string): WebhookConfig | undefined {
    return this.webhooks.get(id);
  }

  getWebhookDeliveries(webhookId: string, limit: number): WebhookDelivery[] {
    return Array.from(this.deliveries.values())
      .filter((delivery) => delivery.webhookId === webhookId)
      .slice(-Math.max(0, limit));
  }

  getDelivery(deliveryId: string): WebhookDelivery | undefined {
    return this.deliveries.get(deliveryId);
  }

  getAnalytics(webhookId: string): WebhookAnalytics {
    const deliveries = this.getWebhookDeliveries(webhookId, Number.MAX_SAFE_INTEGER);
    const totalDeliveries = deliveries.length;
    const successfulDeliveries = deliveries.filter(
      (delivery) => delivery.status === 'delivered'
    ).length;
    const failedDeliveries = deliveries.filter((delivery) => delivery.status === 'failed').length;
    const pendingDeliveries = deliveries.filter((delivery) =>
      ['pending', 'retrying', 'paused'].includes(delivery.status)
    ).length;
    const retryCount = deliveries.reduce(
      (sum, delivery) => sum + Math.max(0, delivery.attempts - 1),
      0
    );
    const avgAttempts = totalDeliveries
      ? deliveries.reduce((sum, d) => sum + d.attempts, 0) / totalDeliveries
      : 0;

    return {
      webhookId,
      totalDeliveries,
      successfulDeliveries,
      failedDeliveries,
      retryCount,
      pendingDeliveries,
      successRate: totalDeliveries ? successfulDeliveries / totalDeliveries : 0,
      avgAttempts,
      lastSuccessAt: deliveries
        .filter((delivery) => delivery.status === 'delivered' && delivery.deliveredAt)
        .map((delivery) => delivery.deliveredAt as number)
        .sort((a, b) => b - a)[0],
      lastFailureAt: deliveries
        .filter((delivery) => delivery.status === 'failed' && delivery.updatedAt)
        .map((delivery) => delivery.updatedAt)
        .sort((a, b) => b - a)[0],
    };
  }

  async checkWebhookHealth(id: string): Promise<WebhookConfig> {
    const webhook = this.webhooks.get(id);
    if (!webhook) throw new Error(`Webhook ${id} not found`);

    const checkedAt = now();
    try {
      const response = await this.fetchImpl(webhook.url, { method: 'HEAD' });
      const healthy = response.ok;
      const next: WebhookConfig = {
        ...webhook,
        lastHealthCheckAt: checkedAt,
        lastHealthStatus: healthy ? 'healthy' : 'unhealthy',
        updatedAt: checkedAt,
      };
      this.webhooks.set(id, next);
      return next;
    } catch {
      const next: WebhookConfig = {
        ...webhook,
        lastHealthCheckAt: checkedAt,
        lastHealthStatus: 'unhealthy',
        updatedAt: checkedAt,
      };
      this.webhooks.set(id, next);
      return next;
    }
  }

  async deliverEvent(input: WebhookEventInput): Promise<WebhookDeliveryResult | null> {
    const webhook = this.webhooks.get(input.webhookId);
    if (!webhook || webhook.merchantId !== input.merchantId) return null;
    if (!isWebhookEventAllowed(webhook, input.eventType)) return null;

    const payload = buildWebhookPayload(input);
    const signature = signWebhookPayload(payload, webhook.secretKey);
    const idempotencyKey = `${payload.id}:${webhook.id}`;
    if (this.deliveredKeys.has(idempotencyKey)) {
      const delivery: WebhookDelivery = {
        id: createId('del'),
        webhookId: webhook.id,
        eventId: payload.id,
        eventType: payload.eventType,
        url: webhook.url,
        payload,
        status: 'skipped',
        attempts: 0,
        maxAttempts: webhook.retryPolicy.maxRetries,
        createdAt: now(),
        updatedAt: now(),
        signature,
        idempotencyKey,
      };
      this.deliveries.set(delivery.id, delivery);
      return { delivery };
    }

    const delivery: WebhookDelivery = {
      id: createId('del'),
      webhookId: webhook.id,
      eventId: payload.id,
      eventType: payload.eventType,
      url: webhook.url,
      payload,
      status: 'pending',
      attempts: 0,
      maxAttempts: webhook.retryPolicy.maxRetries,
      createdAt: now(),
      updatedAt: now(),
      signature,
      idempotencyKey,
    };

    this.deliveries.set(delivery.id, delivery);
    const result = await this.sendWithRetry(webhook, delivery);
    this.deliveries.set(delivery.id, result.delivery);

    if (result.delivery.status === 'delivered') {
      this.deliveredKeys.add(idempotencyKey);
    }
    return result;
  }

  async retryWebhookDelivery(deliveryId: string): Promise<WebhookDeliveryResult> {
    const existing = this.deliveries.get(deliveryId);
    if (!existing) throw new Error(`Delivery ${deliveryId} not found`);
    const webhook = this.webhooks.get(existing.webhookId);
    if (!webhook) throw new Error(`Webhook ${existing.webhookId} not found`);

    const restarted: WebhookDelivery = {
      ...existing,
      attempts: 0,
      status: 'retrying',
      updatedAt: now(),
      nextRetryAt: undefined,
      errorMessage: undefined,
    };

    this.deliveries.set(deliveryId, restarted);
    const result = await this.sendWithRetry(webhook, restarted);
    this.deliveries.set(deliveryId, result.delivery);

    if (result.delivery.status === 'delivered') {
      this.deliveredKeys.add(existing.idempotencyKey);
    }
    return result;
  }

  private async sendWithRetry(
    webhook: WebhookConfig,
    delivery: WebhookDelivery
  ): Promise<WebhookDeliveryResult> {
    const payloadBody = JSON.stringify(delivery.payload);
    if (Buffer.byteLength(payloadBody, 'utf8') > MAX_PAYLOAD_BYTES) {
      return this.finalizeDelivery(webhook, delivery, {
        status: 'failed',
        errorMessage: 'Payload exceeds 1MB limit',
      });
    }

    const headers = {
      'Content-Type': 'application/json',
      'X-SubTrackr-Signature': delivery.signature,
      'X-SubTrackr-Event-Type': delivery.eventType,
      'X-SubTrackr-Event-Id': delivery.eventId,
      'X-SubTrackr-Idempotency-Key': delivery.idempotencyKey,
    };

    let attempt = delivery.attempts;
    let lastError: string | undefined;
    const maxAttempts = Math.max(1, webhook.retryPolicy.maxRetries + 1);

    while (attempt < maxAttempts) {
      attempt += 1;
      const attemptAt = now();
      const next: WebhookDelivery = {
        ...delivery,
        status: attempt === 1 ? 'pending' : 'retrying',
        attempts: attempt,
        lastAttemptAt: attemptAt,
        updatedAt: attemptAt,
      };
      this.deliveries.set(delivery.id, next);

      try {
        const response = await this.fetchImpl(webhook.url, {
          method: 'POST',
          headers,
          body: payloadBody,
        });

        if (!response.ok) {
          throw new Error(`HTTP ${response.status}`);
        }

        return this.finalizeDelivery(
          webhook,
          next,
          {
            status: 'delivered',
            responseCode: response.status,
            deliveredAt: now(),
          },
          response
        );
      } catch (error) {
        lastError = error instanceof Error ? error.message : 'Webhook delivery failed';
        const isLastAttempt = attempt >= maxAttempts;
        const delay = this.computeDelay(webhook.retryPolicy, attempt);

        if (isLastAttempt) {
          return this.finalizeDelivery(webhook, next, {
            status: 'failed',
            errorMessage: lastError,
            responseCode: undefined,
          });
        }

        const retried: WebhookDelivery = {
          ...next,
          status: 'retrying',
          errorMessage: lastError,
          nextRetryAt: now() + delay,
        };
        this.deliveries.set(delivery.id, retried);
        await this.sleepImpl(delay);
      }
    }

    return this.finalizeDelivery(webhook, delivery, {
      status: 'failed',
      errorMessage: lastError ?? 'Webhook delivery failed',
    });
  }

  private finalizeDelivery(
    webhook: WebhookConfig,
    delivery: WebhookDelivery,
    patch: Partial<WebhookDelivery> & { status: WebhookDeliveryStatus },
    response?: Response
  ): WebhookDeliveryResult {
    const next: WebhookDelivery = {
      ...delivery,
      ...patch,
      updatedAt: now(),
    };
    this.deliveries.set(delivery.id, next);

    const configPatch: Partial<WebhookConfig> = {
      updatedAt: next.updatedAt,
      successCount: next.status === 'delivered' ? webhook.successCount + 1 : webhook.successCount,
      failureCount: next.status === 'failed' ? webhook.failureCount + 1 : webhook.failureCount,
      lastHealthStatus:
        next.status === 'delivered'
          ? 'healthy'
          : next.status === 'failed'
            ? 'degraded'
            : webhook.lastHealthStatus,
    };
    this.webhooks.set(webhook.id, { ...webhook, ...configPatch });

    return { delivery: next, response };
  }

  private computeDelay(policy: WebhookRetryPolicy, attempt: number): number {
    const factor = policy.backoffFactor ?? DEFAULT_RETRY_POLICY.backoffFactor ?? 2;
    const rawDelay = Math.floor(policy.initialDelayMs * Math.pow(factor, Math.max(0, attempt - 1)));
    return Math.min(rawDelay, policy.maxDelayMs);
  }
}

export const webhookDeliveryService = new WebhookDeliveryService();

File: contracts/subscription/src/webhook.rs
use soroban_sdk::{Address, Env, String, Vec};
use subtrackr_types::{
    StorageKey, Subscription, SubscriptionStatus, WebhookConfig, WebhookDelivery,
    WebhookDeliveryStatus, WebhookEventType, WebhookRetryPolicy,
};

use crate::{
    events::build_payload, storage_instance_get, storage_instance_set, storage_persistent_get,
    storage_persistent_remove, storage_persistent_set,
};

fn webhook_ids_for_merchant(env: &Env, storage: &Address, merchant: &Address) -> Vec<u64> {
    storage_persistent_get(env, storage, StorageKey::MerchantWebhooks(merchant.clone()))
        .unwrap_or(Vec::new(env))
}

fn set_webhook_ids_for_merchant(env: &Env, storage: &Address, merchant: &Address, ids: Vec<u64>) {
    storage_persistent_set(env, storage, StorageKey::MerchantWebhooks(merchant.clone()), ids);
}

fn deliveries_for_webhook(env: &Env, storage: &Address, webhook_id: u64) -> Vec<u64> {
    storage_persistent_get(env, storage, StorageKey::WebhookDeliveriesByWebhook(webhook_id))
        .unwrap_or(Vec::new(env))
}

fn set_deliveries_for_webhook(env: &Env, storage: &Address, webhook_id: u64, ids: Vec<u64>) {
    storage_persistent_set(
        env,
        storage,
        StorageKey::WebhookDeliveriesByWebhook(webhook_id),
        ids,
    );
}

fn webhook_supports_event(config: &WebhookConfig, event_type: &WebhookEventType) -> bool {
    if config.is_paused {
        return false;
    }
    for configured in config.events.iter() {
        if configured == *event_type {
            return true;
        }
    }
    false
}

fn next_webhook_id(env: &Env, storage: &Address) -> u64 {
    let mut count: u64 = storage_instance_get(env, storage, StorageKey::WebhookCount).unwrap_or(0);
    count += 1;
    storage_instance_set(env, storage, StorageKey::WebhookCount, count);
    count
}

fn next_delivery_id(env: &Env, storage: &Address) -> u64 {
    let mut count: u64 =
        storage_instance_get(env, storage, StorageKey::WebhookDeliveryCount).unwrap_or(0);
    count += 1;
    storage_instance_set(env, storage, StorageKey::WebhookDeliveryCount, count);
    count
}

fn compute_delay(policy: &WebhookRetryPolicy, attempt: u32) -> u64 {
    let mut delay = policy.initial_delay_secs;
    let mut i = 1u32;
    while i < attempt {
        delay = delay.saturating_mul(policy.backoff_factor as u64);
        i += 1;
    }
    if delay > policy.max_delay_secs {
        policy.max_delay_secs
    } else {
        delay
    }
}

pub(crate) fn emit_subscription_event(
    env: &Env,
    storage: &Address,
    merchant: &Address,
    event_type: WebhookEventType,
    subscription: &Subscription,
    plan: &subtrackr_types::Plan,
    previous_status: SubscriptionStatus,
) {
    let webhook_ids = webhook_ids_for_merchant(env, storage, merchant);
    let mut i = 0u32;
    while i < webhook_ids.len() {
        let webhook_id = webhook_ids.get_unchecked(i);
        if let Some(config) = storage_persistent_get::<WebhookConfig>(
            env,
            storage,
            StorageKey::Webhook(webhook_id),
        ) {
            if !webhook_supports_event(&config, &event_type) {
                i += 1;
                continue;
            }

            let delivery_id = next_delivery_id(env, storage);
            let payload = build_payload(
                env,
                webhook_id,
                event_type.clone(),
                merchant,
                subscription,
                plan,
                previous_status.clone(),
            );
            let delivery = WebhookDelivery {
                id: delivery_id,
                webhook_id,
                event_id: payload.id,
                event_type,
                payload,
                status: if config.is_paused {
                    WebhookDeliveryStatus::Paused
                } else {
                    WebhookDeliveryStatus::Pending
                },
                attempts: 0,
                max_attempts: config.retry_policy.max_retries,
                next_retry_at: 0,
                last_attempt_at: 0,
                delivered_at: 0,
                response_code: 0,
                error_message: String::from_str(env, ""),
                signature: String::from_str(env, ""),
                created_at: env.ledger().timestamp(),
                updated_at: env.ledger().timestamp(),
            };
            storage_persistent_set(env, storage, StorageKey::WebhookDelivery(delivery_id), delivery);

            let mut deliveries = deliveries_for_webhook(env, storage, webhook_id);
            deliveries.push_back(delivery_id);
            set_deliveries_for_webhook(env, storage, webhook_id, deliveries);
        }
        i += 1;
    }
}

#[soroban_sdk::contractimpl]
impl super::SubTrackrSubscription {
    pub fn register_webhook(
        env: Env,
        proxy: Address,
        storage: Address,
        mut config: WebhookConfig,
    ) -> u64 {
        proxy.require_auth();
        config.merchant.require_auth();

        let id = next_webhook_id(&env, &storage);
        config.id = id;
        config.created_at = env.ledger().timestamp();
        config.updated_at = config.created_at;
        config.health_check_at = 0;
        config.healthy = true;
        config.success_count = 0;
        config.failure_count = 0;

        storage_persistent_set(&env, &storage, StorageKey::Webhook(id), config.clone());

        let mut ids = webhook_ids_for_merchant(&env, &storage, &config.merchant);
        ids.push_back(id);
        set_webhook_ids_for_merchant(&env, &storage, &config.merchant, ids);
        id
    }

    pub fn update_webhook(
        env: Env,
        proxy: Address,
        storage: Address,
        id: u64,
        mut config: WebhookConfig,
    ) {
        proxy.require_auth();
        config.merchant.require_auth();

        let current: WebhookConfig =
            storage_persistent_get(&env, &storage, StorageKey::Webhook(id))
                .expect("Webhook not found");

        assert!(current.merchant == config.merchant, "Webhook merchant mismatch");
        config.id = id;
        config.created_at = current.created_at;
        config.updated_at = env.ledger().timestamp();
        config.success_count = current.success_count;
        config.failure_count = current.failure_count;
        config.health_check_at = current.health_check_at;
        config.healthy = current.healthy;

        storage_persistent_set(&env, &storage, StorageKey::Webhook(id), config);
    }

    pub fn delete_webhook(env: Env, proxy: Address, storage: Address, id: u64) {
        proxy.require_auth();
        let config: WebhookConfig = storage_persistent_get(&env, &storage, StorageKey::Webhook(id))
            .expect("Webhook not found");
        config.merchant.require_auth();

        let ids = webhook_ids_for_merchant(&env, &storage, &config.merchant);
        let mut next_ids = Vec::new(&env);
        for existing in ids.iter() {
            if existing != id {
                next_ids.push_back(existing);
            }
        }
        set_webhook_ids_for_merchant(&env, &storage, &config.merchant, next_ids);
        storage_persistent_remove(
            &env,
            &storage,
            StorageKey::WebhookDeliveriesByWebhook(id),
        );
        storage_persistent_remove(&env, &storage, StorageKey::Webhook(id));
    }

    pub fn pause_webhook(env: Env, proxy: Address, storage: Address, id: u64) {
        proxy.require_auth();
        let mut config: WebhookConfig =
            storage_persistent_get(&env, &storage, StorageKey::Webhook(id))
                .expect("Webhook not found");
        config.merchant.require_auth();
        config.is_paused = true;
        config.updated_at = env.ledger().timestamp();
        storage_persistent_set(&env, &storage, StorageKey::Webhook(id), config);
    }

    pub fn resume_webhook(env: Env, proxy: Address, storage: Address, id: u64) {
        proxy.require_auth();
        let mut config: WebhookConfig =
            storage_persistent_get(&env, &storage, StorageKey::Webhook(id))
                .expect("Webhook not found");
        config.merchant.require_auth();
        config.is_paused = false;
        config.updated_at = env.ledger().timestamp();
        storage_persistent_set(&env, &storage, StorageKey::Webhook(id), config);
    }

    pub fn list_webhooks(env: Env, proxy: Address, storage: Address, merchant: Address) -> Vec<WebhookConfig> {
        proxy.require_auth();
        let ids = webhook_ids_for_merchant(&env, &storage, &merchant);
        let mut items = Vec::new(&env);
        for webhook_id in ids.iter() {
            if let Some(config) =
                storage_persistent_get::<WebhookConfig>(&env, &storage, StorageKey::Webhook(webhook_id))
            {
                items.push_back(config);
            }
        }
        items
    }

    pub fn get_webhook_deliveries(
        env: Env,
        proxy: Address,
        storage: Address,
        webhook_id: u64,
        limit: u32,
    ) -> Vec<WebhookDelivery> {
        proxy.require_auth();
        let ids = deliveries_for_webhook(&env, &storage, webhook_id);
        let mut items = Vec::new(&env);
        let mut i = 0u32;
        while i < ids.len() && i < limit {
            let delivery_id = ids.get_unchecked(i);
            if let Some(delivery) = storage_persistent_get::<WebhookDelivery>(
                &env,
                &storage,
                StorageKey::WebhookDelivery(delivery_id),
            ) {
                items.push_back(delivery);
            }
            i += 1;
        }
        items
    }

    pub fn retry_webhook_delivery(env: Env, proxy: Address, storage: Address, delivery_id: u64) {
        proxy.require_auth();
        let mut delivery: WebhookDelivery =
            storage_persistent_get(&env, &storage, StorageKey::WebhookDelivery(delivery_id))
                .expect("Webhook delivery not found");

        let config: WebhookConfig = storage_persistent_get(
            &env,
            &storage,
            StorageKey::Webhook(delivery.webhook_id),
        )
        .expect("Webhook not found");
        config.merchant.require_auth();

        if delivery.attempts > config.retry_policy.max_retries {
            delivery.status = WebhookDeliveryStatus::Failed;
        } else {
            delivery.attempts += 1;
            delivery.status = WebhookDeliveryStatus::Retrying;
            delivery.last_attempt_at = env.ledger().timestamp();
            delivery.next_retry_at = env.ledger().timestamp()
                + compute_delay(&config.retry_policy, delivery.attempts);
        }
        delivery.updated_at = env.ledger().timestamp();
        storage_persistent_set(&env, &storage, StorageKey::WebhookDelivery(delivery_id), delivery);
    }

    pub fn get_webhook_health(
        env: Env,
        proxy: Address,
        storage: Address,
        webhook_id: u64,
    ) -> WebhookConfig {
        proxy.require_auth();
        let mut config: WebhookConfig =
            storage_persistent_get(&env, &storage, StorageKey::Webhook(webhook_id))
                .expect("Webhook not found");
        config.merchant.require_auth();

        let deliveries = deliveries_for_webhook(&env, &storage, webhook_id);
        let mut failures = 0u64;
        let mut successes = 0u64;
        for delivery_id in deliveries.iter() {
            if let Some(delivery) = storage_persistent_get::<WebhookDelivery>(
                &env,
                &storage,
                StorageKey::WebhookDelivery(delivery_id),
            ) {
                match delivery.status {
                    WebhookDeliveryStatus::Delivered => successes += 1,
                    WebhookDeliveryStatus::Failed => failures += 1,
                    _ => {}
                }
            }
        }

        config.healthy = failures <= successes;
        config.health_check_at = env.ledger().timestamp();
        config.updated_at = config.health_check_at;
        storage_persistent_set(&env, &storage, StorageKey::Webhook(webhook_id), config.clone());
        config
    }
}
